### PR TITLE
Chameleon gene fix

### PR DESCRIPTION
**Pull request by flappybatpal**
[BALANCE][BUG]

## About the PR 
Bug fix for https://github.com/goonstation/goonstation/issues/2096

## Why's this needed? 
Chameleon was comparing world.timeofday to owner.l_move_time which didn't make sense. From what I can tell this was totally broken.
I've changed this around to keeping track of the world time. The logic may not be perfect but it behaved in testing.
We should improve the way the cloak is dropped as right now you could stand still and fire a SPES 12 invisibly!

## Changelog
```
(u)Flappybat:
(*)Fixed chameleon gene not disabling on movement
(*)Chameleon invisibility drops when attacked, bringing closer parity to cloaking device
```

### CODE REVIEW

**Comment by Sovexe** [comment]
I feel like this might be better off registering the owner for `COMSIG_MOVABLE_MOVED` or something on `cast()`

That way if you walk or get bumped or whatever it just de-cloaks you, rather than checking if you moved and updating overlays and setting invisibility over and over

**Comment by flappybatpal** [comment]
Having trouble with the signals. Movement is fine but I want to deactivate on attack/being attacked to be a closer match to the traitor cloak. Think it can't hear the attack signals from the datum, not sure how to proceed.

**Comment by Sovexe** [comment]
We have a signal for being attacked not for attacking. Cloak ends on attacking because there are checks throughout the code that check if you have an active `cloaking_device` and then disable it

**Comment by flappybatpal** [comment]
Changed to TIME and ditched COMSIG_ATTACKBY.

Can't work out how to get attacking to work. Would need to call the proc somehow from other DMs but I can't work out how to do it without nulls. Got an accidental merge in too with 2bf87fb and can't work out how to get rid of that either.

I'm leaving this alone now, it works as far as fixing the movement bug and slightly rebalancing chameleon to match invisibility so your cloak drops if you are attacked instead of being perma stealth until you are moved.

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 1]
```diff
@@ -1902,6 +1902,7 @@
 	lockedTries = 8
 	stability_loss = 20
 	cooldown = 0
+	var/last_moved = 0;
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1905]:
```suggestion
	var/last_moved = 0
```

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 2]
```diff
@@ -1921,14 +1922,20 @@
 
 	OnLife()
 		if(..()) return
+		if(!src.active) return
 		if(isliving(owner))
 			var/mob/living/L = owner
-			if ((world.timeofday - owner.l_move_time) >= 30 && can_act(owner) && src.active)
+			if ((TIME - last_moved) >= 30 && can_act(owner))
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1928]:
```suggestion
			if (TIME - last_moved >= 3 SECONDS && can_act(owner))
```

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 3]
```diff
@@ -1921,14 +1922,20 @@
...
 				L.UpdateOverlays(overlay_image, id)
 				L.invisibility = 1
-			else
-				L.UpdateOverlays(null, id)
-				L.invisibility = 0
+
+	proc/decloak()
+		boutput(world, "decloak")
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1933]:
```suggestion
```
*scream

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 4]
```diff
@@ -1944,9 +1951,13 @@
 		if (CH.active)
 			boutput(usr, "You stop using your chameleon cloaking.")
 			CH.active = 0
+			CH.UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE))
+			CH.decloak()
 		else
 			boutput(usr, "You start using your chameleon cloaking.")
+			CH.last_moved = TIME
 			CH.active = 1
+			CH.RegisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE), /datum/bioEffect/power/chameleon/proc/decloak, FALSE)
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1960]:
```suggestion
			CH.RegisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE), /datum/bioEffect/power/chameleon/proc/decloak)
```
FALSE is already the default

File: code/modules/medical/genetics/bioEffects/powers.dm (hunk 3 above)
Comment (flappybatpal) [reply code/modules/medical/genetics/bioEffects/powers.dm:1933]: Jeez I was sure I removed this.

**Review by pali6 — CHANGES_REQUESTED** [review]
Actually one more thing
`OnRemove()` should also unregister the signals if they are registered.

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 5]
```diff
@@ -1917,18 +1918,25 @@
 			var/mob/living/L = owner
 			L.UpdateOverlays(null, id)
 			L.invisibility = 0
+		if (src.active)
+			CH.UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE))
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1922]:
```suggestion
			linked_power.UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE))
```

File: code/modules/medical/genetics/bioEffects/powers.dm [hunk 6]
```diff
@@ -1917,18 +1918,25 @@
 			var/mob/living/L = owner
 			L.UpdateOverlays(null, id)
 			L.invisibility = 0
+		if (src.active)
+			linked_power.UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE))
```
Comment (pali6) [inline code/modules/medical/genetics/bioEffects/powers.dm:1922]:
```suggestion
			src.UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_MOB_ATTACKED_PRE))
```

**Review by pali6 — APPROVED** [review]

**Pull request merged**